### PR TITLE
Simulated annealing for SDP

### PR DESCRIPTION
**Pull request by panagiotisrep**
The improved version of the simulated annealing algorithm for convex optimization from Kalai and Vempala, “Simulated Annealing for Convex Optimization”, for spectrahedra (semidefinite programming). This PR:  

-- Implements the exact Hamiltonian Monte Carlo for sampling from the exponential distribution.
-- Employs `Eigen`, `Spectra` and `arpackpp` to implement the boundary oracle (which reduces to a Quadratic Eigenvalue Problem).
-- Uses the deterministic cooling scheme of [Kalai and Vempala, “Simulated Annealing for Convex Optimization”].
-- Does not applies the isotropization step given in [Kalai and Vempala, “Simulated Annealing for Convex Optimization”].

Some possible next steps based on this PR:

-- Fix numerical issues close to the boundary.  
-- Introduce an adaptive version of Hamiltonian Monte Carlo to reduce the average number of reflections.
-- Exploit other distributions to select a point on the trajectory in each step of the random walk.
-- Introduce a more sophisticated cooling scheme.
-- Test various isotropization techniques.  

More info on implementation / testing [here](https://www.dropbox.com/s/65vrb3828t82ub2/Randomized%20cutting%20plane.pdf?dl=0).

### CODE REVIEW

File: R-proj/src/Makevars [hunk 1]
```diff
@@ -1,4 +1,4 @@
-PKG_CPPFLAGS=-I../../external/boost -I../../external/LPsolve_src/run_headers -I../../external/minimum_ellipsoid -I../../include 
+PKG_CPPFLAGS=-I../../external/boost -I../../external/LPsolve_src/run_headers -I../../external/minimum_ellipsoid -I../../include -I../../include/convex_bodies/spectrahedra
```
Comment (TolisChal) [inline R-proj/src/Makevars]: please do not include any subdirectory of `/root/include in makevars.

File: R-proj/src/Makevars.win [hunk 2]
```diff
@@ -1,4 +1,4 @@
-PKG_CPPFLAGS=-I../../external/boost -I../../external/LPsolve_src/run_headers -I../../external/minimum_ellipsoid -I../../include 
+PKG_CPPFLAGS=-I../../external/boost -I../../external/LPsolve_src/run_headers -I../../external/minimum_ellipsoid -I../../include -I../../include/convex_bodies/spectrahedra
```
Comment (TolisChal) [inline R-proj/src/Makevars.win]: same as before

File: examples/spectrahedra/README.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
+# Examples for Spectrahedra
```
Comment (TolisChal) [inline examples/spectrahedra/README.md:1]: Please pull changes from develop branch. The Readme file is much more short now.

File: R-proj/src/Makevars (hunk 1 above)
Comment (panagiotisrep) [reply R-proj/src/Makevars]: this will change when i merge some changes from previous pull request

File: R-proj/src/Makevars.win (hunk 2 above)
Comment (panagiotisrep) [reply R-proj/src/Makevars.win]: this will change when i merge some changes from previous pull request

**Review by vissarion — APPROVED** [review]
Thanks for this PR @panagiotisrep !

**Review by TolisChal — APPROVED** [review]

**Pull request merged**